ucp/useraccounts_module.php

```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
 					{
-						// Reject duplicates.
-						$sql = "SELECT COUNT(*) as duplcount
-							FROM $user_badge_table
-							WHERE badge_id = $markedbadge
-							AND user_id = $this_user_id";
+						// Query valid badges
+						$sql = "SELECT badge_id
+							FROM $badge_table
+							WHERE badge_default = TRUE";
 						if (!($result = $this->db->sql_query($sql)))
 						{
-							trigger_error($this->user->lang('UCP_DUPLICATE_USER_BADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+							trigger_error($this->user->lang('UCP_DEFAULT_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 						}
-						$duplicates = (int) $this->db->sql_fetchfield('duplcount');
+						$default_badges = $this->db->sql_fetchrowset($result);
 						$this->db->sql_freeresult($result);
-						if (!$duplicates)
+
+						$sql = "SELECT badge_id
+							FROM $badge_avail_table
+							WHERE user_id = $this_user_id";
+						if (!($result = $this->db->sql_query($sql)))
 						{
-							// Add a new user badge.
-							$data = array(
-								'user_id' => $this_user_id,
-								'badge_id' => $markedbadge,
-								'badge_order' => null // Order value will be validated after loop.
-							);
-							$sql = 'INSERT INTO ' . $user_badge_table . $this->db->sql_build_array('INSERT', $data);
-							$this->db->sql_query($sql);
+							trigger_error($this->user->lang('UCP_AVAIL_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 						}
+						$custom_badges = $this->db->sql_fetchrowset($result);
+						$this->db->sql_freeresult($result);
+
+						// Prepare an array for inserting badges
+						$sql_array = array();
+						foreach ($markedbadges as $markedbadge)
+						{
+							// Validate requested badges and reject duplicates.
+							if (($this->core->in_2d_array($markedbadge, $default_badges) || $this->core->in_2d_array($markedbadge, $custom_badges)) && !$this->core->in_2d_array($markedbadge, $user_badges))
+							{
+								$sql_array[] = array(
+									'user_id' => $this_user_id,
+									'badge_id' => $markedbadge,
+									'badge_order' => null // Order value will be validated after loop.
+								);
+							}
+						}
+
+						if(!empty($sql_array))
+						{
+							// No need to escape strings or force variable types when using sql_multi_insert().
+							if (!$this->db->sql_multi_insert($user_badge_table, $sql_array))
+							{
+								trigger_error($this->user->lang('UCP_INSERT_MARKED_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+							}
+						}
+
+						if (!$this->core->validate_badge_order($this_user_id))
+						{
+							trigger_error($this->user->lang('UCP_BADGE_ORDER_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
```

Comment:
Do we really need to check this?  `validate_badge_order` will only return false if the initial badge query fails, but I think phpBB will generate its own error page on most of these cases?